src/script/tari_script.rs

```diff
@@ -0,0 +1,345 @@
+        let b = Blake256::new()
+            .chain(pub_key.as_bytes())
+            .chain(&self.as_bytes())
+            .result()
+            .to_vec();
+        RistrettoSecretKey::from_bytes(&b).map_err(|_| ScriptError::InvalidSignature)
+    }
+
+    fn execute_opcode(&self, opcode: &Opcode, stack: &mut ExecutionStack) -> Result<(), ScriptError> {
+        use StackItem::*;
+        match opcode {
+            Opcode::Return => Err(ScriptError::Return),
+            Opcode::Add => TariScript::handle_op_add(stack),
+            Opcode::Sub => TariScript::handle_op_sub(stack),
+            Opcode::CheckSig => match self.check_sig(stack)? {
+                true => stack.push(Number(0)),
+                false => stack.push(Number(1)),
+            },
+            Opcode::CheckSigVerify => match self.check_sig(stack)? {
+                true => stack.push(Number(0)),
+                false => Err(ScriptError::InvalidSignature),
+            },
+            Opcode::RevRot => stack.push_down(2),
+        }
+    }
+
+    fn handle_op_add(stack: &mut ExecutionStack) -> Result<(), ScriptError> {
+        use StackItem::*;
+        let top = stack.pop().ok_or(ScriptError::StackUnderflow)?;
+        let two = stack.pop().ok_or(ScriptError::StackUnderflow)?;
+        match (top, two) {
+            (Number(v1), Number(v2)) => stack.push(Number(v1.checked_add(v2).ok_or(ScriptError::ValueExceedsBounds)?)),
+            (Commitment(c1), Commitment(c2)) => stack.push(Commitment(&c1 + &c2)),
+            (PublicKey(p1), PublicKey(p2)) => stack.push(PublicKey(&p1 + &p2)),
+            (Signature(s1), Signature(s2)) => stack.push(Signature(&s1 + &s2)),
+            (_, _) => Err(ScriptError::IncompatibleTypes),
+        }
+    }
+
+    fn handle_op_sub(stack: &mut ExecutionStack) -> Result<(), ScriptError> {
+        use StackItem::*;
+        let top = stack.pop().ok_or(ScriptError::StackUnderflow)?;
+        let two = stack.pop().ok_or(ScriptError::StackUnderflow)?;
+        match (top, two) {
+            (Number(v1), Number(v2)) => stack.push(Number(v2.checked_sub(v1).ok_or(ScriptError::ValueExceedsBounds)?)),
+            (Commitment(c1), Commitment(c2)) => stack.push(Commitment(&c2 - &c1)),
+            (..) => Err(ScriptError::IncompatibleTypes),
+        }
+    }
+
+    fn check_sig(&self, stack: &mut ExecutionStack) -> Result<bool, ScriptError> {
+        use StackItem::*;
+        let pk = stack.pop().ok_or(ScriptError::StackUnderflow)?;
+        let sig = stack.pop().ok_or(ScriptError::StackUnderflow)?;
+        match (pk, sig) {
+            (PublicKey(p), Signature(s)) => {
+                let m = self.script_message(&p)?;
+                Ok(s.verify(&p, &m))
+            },
+            (..) => Err(ScriptError::IncompatibleTypes),
+        }
+    }
+
+    fn is_successful(stack: &ExecutionStack) -> Result<(), ScriptError> {
+        if stack.size() != 1 {
```

Comment:
It might also be helpful in debugging to throw an EmptyStack error here if there's nothing

Reply:
Maybe we could return the stack height on error, but I don't see the value in a separate error type